src/develop/imageop.c

```diff
@@ -2746,6 +2746,48 @@ int dt_iop_count_instances(dt_iop_module_so_t *module)
   return inst_count;
 }
 
+static gboolean _postponed_history_update(gpointer data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t*)data;
+  dt_dev_add_history_item(darktable.develop, self, TRUE);
+  self->timeout_handle = 0;
+  return FALSE; //cancel the timer
+}
+
+/** queue a delayed call of the add_history function after user interaction, to capture parameter updates (but not */
+/** too often). */
+void dt_iop_queue_history_update(dt_iop_module_t *module, gboolean extend_prior)
+{
+  if (module->timeout_handle)
+  {
+    if (extend_prior)
+    {
+      // we already queued an update, but we don't want to have the update happen until the timeout expires
+      // without any activity, so cancel the queued callback
+      g_source_remove(module->timeout_handle);
+    }
+    else
+    {
+      // let the existing event happen as scheduled
+      return;
+    }
+  }
+  // adaptively set the timeout to 150% of the average time the past several pixelpipe runs took, clamped
+  //   to keep updates from appearing to be too sluggish (though early iops such as rawdenoise may have
+  //   multiple very slow iops following them, leading to >1000ms processing times)
+  const int delay = CLAMP(darktable.develop->average_delay * 3 / 2, 10, 1200);
+  module->timeout_handle = g_timeout_add(delay, _postponed_history_update, module);
+}
```

Comment:
This function is never called with `extend_prior == TRUE`. Thus, this function can be simplified, removing the 2nd argument . Even if one considers the 2nd argument as necessary, this function should be rewritten for clarity. In particular the early `return` in the `else` branch obfuscates what goes on here.

Reply:
Changed.  The `extend_prior` parameter is for future-proofing, if some other iop wants to use this mechanism that way.